lib/rucio/core/rule.py

```diff
@@ -427,33 +472,55 @@ def add_rules(dids, rules, *, session: "Session", logger=logging.log):
-                    .filter(models.RSEFileAssociation.scope == did.scope,
-                            models.RSEFileAssociation.name == did.name,
-                            models.RSEFileAssociation.state == ReplicaState.AVAILABLE,
-                            models.RSE.rse_type != RSEType.TAPE).count()
+                stmt = select(
+                    func.count(models.RSEFileAssociation.rse_id)
+                ).join(
+                    models.RSE,
+                    models.RSE.id == models.RSEFileAssociation.rse_id
+                ).where(
+                    and_(models.RSEFileAssociation.scope == did.scope,
+                         models.RSEFileAssociation.name == did.name,
+                         models.RSEFileAssociation.state == ReplicaState.AVAILABLE,
+                         models.RSE.rse_type != RSEType.TAPE)
+                )
+                replica_cnt = session.execute(stmt).scalar()
                 if replica_cnt == 0:  # Put the rule on the archive
-                    archive = session.query(models.ConstituentAssociation).join(models.RSEFileAssociation,
-                                                                                and_(models.ConstituentAssociation.scope == models.RSEFileAssociation.scope,
-                                                                                     models.ConstituentAssociation.name == models.RSEFileAssociation.name))\
-                        .filter(models.ConstituentAssociation.child_scope == did.scope,
-                                models.ConstituentAssociation.child_name == did.name).first()
+                    stmt = select(
+                        models.ConstituentAssociation
+                    ).join(
+                        models.RSEFileAssociation,
+                        and_(models.ConstituentAssociation.scope == models.RSEFileAssociation.scope,
+                             models.ConstituentAssociation.name == models.RSEFileAssociation.name)
+                    ).where(
+                        and_(models.ConstituentAssociation.child_scope == did.scope,
+                             models.ConstituentAssociation.child_name == did.name)
+                    )
+                    archive = session.execute(stmt).scalars().first()
                     if archive is not None:
                         elem['name'] = archive.name
                         elem['scope'] = archive.scope
                         try:
-                            did = session.query(models.DataIdentifier).filter(models.DataIdentifier.scope == elem['scope'],
-                                                                              models.DataIdentifier.name == elem['name']).one()
+                            stmt = select(
+                                models.DataIdentifier
+                            ).where(
+                                and_(models.DataIdentifier.scope == elem['scopea'],
```

Comment:
`scopea` seems to be a typo, should be `scope`

Reply:
pretty sure this is a vim typo.. changed to just `scope`